tokio/src/sync/mutex.rs

```diff
@@ -352,22 +369,127 @@ where
 
 // === impl MutexGuard ===
 
+impl<'a, T> MutexGuard<'a, T> {
+    /// Makes a new [`MutexGuard`] for a component of the locked data.
+    ///
+    /// This can be used to hold a subfield of the protected data.
+    ///
+    /// This operation cannot fail as the [`MutexGuard`] passed in already locked the mutex.
+    ///
+    /// This is an associated function that needs to be used as `MutexGuard::map(...)`. A method
+    /// would interfere with methods of the same name on the contents of the locked data.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::{Mutex, MutexGuard};
+    ///
+    /// #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+    /// struct Foo(u32);
+    ///
+    /// # #[tokio::main]
+    /// # async fn main() {
+    /// let foo = Mutex::new(Foo(1));
+    ///
+    /// {
+    ///     let mut mapped = MutexGuard::map(foo.lock().await, |f| &mut f.0);
+    ///     *mapped = 2;
+    /// }
+    ///
+    /// assert_eq!(Foo(2), *foo.lock().await);
+    /// # }
+    /// ```
+    ///
+    /// [`MutexGuard`]: struct@MutexGuard
+    #[inline]
+    pub fn map<U, F>(mut this: Self, f: F) -> MutexGuard<'a, U>
+    where
+        F: FnOnce(&mut T) -> &mut U,
+    {
+        let data = f(&mut *this) as *mut U;
+        let s = this.s;
+
+        // Need to forget `this` so that the mutex does not unlock when that is dropped
+        // Essentially transferring ownership to the new `MutexGuard`
+        mem::forget(this);
```

Comment:
We _could_ consider changing `MutexGuard` to own a semaphore permit from the public API `Semaphore` type, rather than using `batch_semaphore` directly. In that case, we wouldn't need to manually release the permit in `MutexGuard`'s drop impl, and the `mem::forget` here would be unnecessary — we can simply move the semaphore permit out of `self` and into the returned mapped guard. 

This would be a larger change, though, so I might consider it as a potential follow-up, rather than something we should do now.